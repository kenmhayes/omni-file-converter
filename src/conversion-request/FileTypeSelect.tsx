import React, { useEffect, useState } from 'react';
import Select, { SelectOption } from './Select';
import {
  DEFAULT_FILE_TYPE, FILE_TYPE_CONVERSION_MAP, FileType, FileTypeConversionMetadata,
} from '../constants/FileTypes';

export interface FileTypeSelectProps {
  onOriginalTypeChange: (type: FileType) => void;
  onConvertTypeChange: (type: FileType) => void;
}

function getFileTypeFromMap(fileTypeKey: string): FileType {
  const metadata: FileTypeConversionMetadata = FILE_TYPE_CONVERSION_MAP[fileTypeKey];
  if (metadata) {
    return metadata.fileType;
  }
  return DEFAULT_FILE_TYPE;
}

function createSelectOptionFromFileType(fileType: FileType): SelectOption {
  return { id: fileType.key, displayValue: fileType.displayValue };
}

function FileTypeSelect(props: FileTypeSelectProps) {
  const [originalTypeOptions, setOriginalTypeOptions] = useState<SelectOption[]>([]);
  const { onOriginalTypeChange, onConvertTypeChange } = props;
  const [convertTypeSelectOptions, setConvertTypeSelectOptions] = useState<SelectOption[]>([]);

  const originalTypeSelectionHandler = (selectedFileType: string) => {
    const newType = getFileTypeFromMap(selectedFileType);
    const conversionData: FileTypeConversionMetadata = FILE_TYPE_CONVERSION_MAP[newType.key];
    setConvertTypeSelectOptions(
      conversionData.supportedConversionTypeKeys.map(
        (key: string) => createSelectOptionFromFileType(getFileTypeFromMap(key)),
      ),
    );

    onOriginalTypeChange(newType);
    onConvertTypeChange(DEFAULT_FILE_TYPE);
  };

  const convertTypeSelectionHandler = (selectedFileType: string) => {
    onConvertTypeChange(getFileTypeFromMap(selectedFileType));
  };

  useEffect(() => {
    const selectOptions: SelectOption[] = [];
    Object.keys(FILE_TYPE_CONVERSION_MAP)
      .forEach((fileTypeKey: string) => {
        if (fileTypeKey !== DEFAULT_FILE_TYPE.key) {
          selectOptions.push(createSelectOptionFromFileType(getFileTypeFromMap(fileTypeKey)));
        }
      });
    setOriginalTypeOptions(selectOptions);
  }, []);

  return (
    <div className="container">
      <div className="row pb-4">
        <Select
          className="col-3"
          options={originalTypeOptions}
          onSelectionChange={originalTypeSelectionHandler}
          placeholderText="Original file type"
        />
      </div>
      <div className="row">
        <Select
          className="col-3"
          options={convertTypeSelectOptions}
          onSelectionChange={convertTypeSelectionHandler}
          placeholderText="Converted file type"
        />
      </div>
    </div>
  );
}

export default FileTypeSelect;
