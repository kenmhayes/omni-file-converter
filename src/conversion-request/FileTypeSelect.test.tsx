import { fireEvent, render } from "@testing-library/react";
import FileTypeSelect from "./FileTypeSelect";
import { act, create } from "react-test-renderer";

const originalSelectText = "Original file type";
const convertedSelectText = "Converted file type";

it('renders a file type select control group', async () => {
    let component: any;
    await act(async () => {
        component = create(
            <FileTypeSelect onOriginalTypeChange={jest.fn()} onConvertTypeChange={jest.fn()} />
        );
    });
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

it('allows for selection of file types', () => {
    const {queryByDisplayValue, getByDisplayValue } = render(
      <FileTypeSelect onOriginalTypeChange={jest.fn()} onConvertTypeChange={jest.fn()} />
    );

    expect(queryByDisplayValue(originalSelectText)).toBeTruthy();
    expect(queryByDisplayValue(convertedSelectText)).toBeTruthy();

    // Cannot select a converted file type without an original designated
    fireEvent.change(getByDisplayValue(convertedSelectText), { target: { value: 'png' }});
    expect(queryByDisplayValue('png')).toBeFalsy();

    // Select jpg in the first select
    fireEvent.change(getByDisplayValue(originalSelectText), { target: { value: 'jpg'}});

    expect(queryByDisplayValue('jpeg/jpg')).toBeTruthy();

    // Select png in the second select
    fireEvent.change(getByDisplayValue(convertedSelectText), { target: { value: 'png' }});

    expect(queryByDisplayValue('png')).toBeTruthy();

    // Select png in the first select, expect empty in second select
    fireEvent.change(getByDisplayValue('jpeg/jpg'), { target: { value: 'png'}});

    expect(queryByDisplayValue(convertedSelectText)).toBeTruthy();

    // Can now select jpg in the second select, as it is a valid conversion for jpg

    fireEvent.change(getByDisplayValue(convertedSelectText), { target: { value: 'jpg'}});

    expect(queryByDisplayValue('jpeg/jpg')).toBeTruthy();
});