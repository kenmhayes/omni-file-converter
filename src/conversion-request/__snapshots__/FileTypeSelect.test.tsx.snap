// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a file type select control group 1`] = `
<div
  className="container"
>
  <div
    className="row pb-4"
  >
    <div
      className="col-3"
    >
      <select
        className="form-select"
        onChange={[Function]}
      >
        <option
          value=""
        >
          Original file type
        </option>
        <option
          value="jpg"
        >
          jpeg/jpg
        </option>
        <option
          value="png"
        >
          png
        </option>
      </select>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-3"
    >
      <select
        className="form-select"
        onChange={[Function]}
      >
        <option
          value=""
        >
          Converted file type
        </option>
      </select>
    </div>
  </div>
</div>
`;
